File: src/baldr/location.cc
#include "baldr/location.h"
#include "baldr/rapidjson_utils.h"
#include "midgard/logging.h"
#include "midgard/pointll.h"
#include "midgard/util.h"

namespace valhalla {
namespace baldr {

Location::SearchFilter::SearchFilter(valhalla::RoadClass min_road_class,
                                     valhalla::RoadClass max_road_class,
                                     bool exclude_tunnel,
                                     bool exclude_bridge,
                                     bool exclude_ramp,
                                     bool exclude_closures)
    : min_road_class_(min_road_class), max_road_class_(max_road_class),
      exclude_tunnel_(exclude_tunnel), exclude_bridge_(exclude_bridge), exclude_ramp_(exclude_ramp),
      exclude_closures_(exclude_closures) {
}

// TODO: get defaults from config singleton
Location::Location(const midgard::PointLL& latlng,
                   const StopType& stoptype,
                   unsigned int min_outbound_reach,
                   unsigned int min_inbound_reach,
                   unsigned long radius,
                   const PreferredSide& side,
                   valhalla::RoadClass street_side_cutoff,
                   const SearchFilter& search_filter,
                   std::optional<int8_t> preferred_layer)
    : latlng_(latlng), stoptype_(stoptype), min_outbound_reach_(min_outbound_reach),
      min_inbound_reach_(min_inbound_reach), radius_(radius), preferred_side_(side),
      node_snap_tolerance_(5), heading_tolerance_(60), search_cutoff_(35000),
      street_side_tolerance_(5), street_side_max_distance_(1000),
      street_side_cutoff_(street_side_cutoff), search_filter_(search_filter),
      preferred_layer_(std::move(preferred_layer)) {
}

bool Location::operator==(const Location& o) const {
  return latlng_ == o.latlng_ && stoptype_ == o.stoptype_ && name_ == o.name_ &&
         street_ == o.street_ && date_time_ == o.date_time_ && heading_ == o.heading_ &&
         heading_tolerance_ == o.heading_tolerance_ &&
         node_snap_tolerance_ == o.node_snap_tolerance_ &&
         street_side_tolerance_ == o.street_side_tolerance_ &&
         street_side_max_distance_ == o.street_side_max_distance_ &&
         street_side_cutoff_ == o.street_side_cutoff_ &&
         min_outbound_reach_ == o.min_outbound_reach_ && min_inbound_reach_ == o.min_inbound_reach_ &&
         radius_ == o.radius_ && preferred_side_ == o.preferred_side_ &&
         display_latlng_ == o.display_latlng_ && preferred_layer_ == o.preferred_layer_;
}

} // namespace baldr
} // namespace valhalla

File: src/baldr/location.cs
using System;
using System.Text.Json;
using valhalla.midgard;

namespace valhalla.baldr
{
    public class Location
    {
        public PointLL LatLng { get; }
        public StopType StopType { get; }
        public string Name { get; set; }
        public string Street { get; set; }
        public string DateTime { get; set; }
        public int? Heading { get; set; }
        public int HeadingTolerance { get; set; }
        public int NodeSnapTolerance { get; set; }
        public int StreetSideTolerance { get; set; }
        public int StreetSideMaxDistance { get; set; }
        public RoadClass StreetSideCutoff { get; set; }
        public int MinOutboundReach { get; set; }
        public int MinInboundReach { get; set; }
        public long Radius { get; set; }
        public PreferredSide PreferredSide { get; set; }
        public PointLL DisplayLatLng { get; set; }
        public int? PreferredLayer { get; set; }
        public SearchFilter SearchFilter { get; }

        public Location(PointLL latlng,
                        StopType stoptype,
                        int minOutboundReach = 0,
                        int minInboundReach = 0,
                        long radius = 0,
                        PreferredSide side = PreferredSide.Either,
                        RoadClass streetSideCutoff = RoadClass.ServiceOther,
                        SearchFilter searchFilter = null,
                        int? preferredLayer = null)
        {
            LatLng = latlng;
            StopType = stoptype;
            MinOutboundReach = minOutboundReach;
            MinInboundReach = minInboundReach;
            Radius = radius;
            PreferredSide = side;
            StreetSideCutoff = streetSideCutoff;
            SearchFilter = searchFilter ?? new SearchFilter();
            PreferredLayer = preferredLayer;
        }

        public bool Equals(Location other)
        {
            return LatLng.Equals(other.LatLng) &&
                   StopType == other.StopType &&
                   Name == other.Name &&
                   Street == other.Street &&
                   DateTime == other.DateTime &&
                   Heading == other.Heading &&
                   HeadingTolerance == other.HeadingTolerance &&
                   NodeSnapTolerance == other.NodeSnapTolerance &&
                   StreetSideTolerance == other.StreetSideTolerance &&
                   StreetSideMaxDistance == other.StreetSideMaxDistance &&
                   StreetSideCutoff == other.StreetSideCutoff &&
                   MinOutboundReach == other.MinOutboundReach &&
                   MinInboundReach == other.MinInboundReach &&
                   Radius == other.Radius &&
                   PreferredSide == other.PreferredSide &&
                   DisplayLatLng.Equals(other.DisplayLatLng) &&
                   PreferredLayer == other.PreferredLayer;
        }

        public JsonElement Json()
        {
            var locationMap = new
            {
                latlng = LatLng,
                stoptype = StopType,
                name = Name,
                street = Street,
                date_time = DateTime,
                heading = Heading,
                heading_tolerance = HeadingTolerance,
                node_snap_tolerance = NodeSnapTolerance,
                street_side_tolerance = StreetSideTolerance,
                street_side_max_distance = StreetSideMaxDistance,
                street_side_cutoff = StreetSideCutoff,
                min_outbound_reach = MinOutboundReach,
                min_inbound_reach = MinInboundReach,
                radius = Radius,
                preferred_side = PreferredSide,
                display_latlng = DisplayLatLng,
                preferred_layer = PreferredLayer
            };
            return JsonDocument.Parse(JsonSerializer.Serialize(locationMap)).RootElement;
        }
    }

    public class SearchFilter
    {
        public RoadClass MinRoadClass { get; }
        public RoadClass MaxRoadClass { get; }
        public bool ExcludeTunnel { get; }
        public bool ExcludeBridge { get; }
        public bool ExcludeToll { get; }
        public bool ExcludeRamp { get; }
        public bool ExcludeFerry { get; }
        public bool ExcludeClosures { get; }
        public float Level { get; }

        public SearchFilter(RoadClass minRoadClass = RoadClass.Motorway,
                            RoadClass maxRoadClass = RoadClass.ServiceOther,
                            bool excludeTunnel = false,
                            bool excludeBridge = false,
                            bool excludeToll = false,
                            bool excludeRamp = false,
                            bool excludeFerry = false,
                            bool excludeClosures = false,
                            float level = 0)
        {
            MinRoadClass = minRoadClass;
            MaxRoadClass = maxRoadClass;
            ExcludeTunnel = excludeTunnel;
            ExcludeBridge = excludeBridge;
            ExcludeToll = excludeToll;
            ExcludeRamp = excludeRamp;
            ExcludeFerry = excludeFerry;
            ExcludeClosures = excludeClosures;
            Level = level;
        }
    }

    public enum StopType
    {
        Break,
        Through,
        Via
    }

    public enum PreferredSide
    {
        Either,
        Same,
        Opposite
    }

    public enum RoadClass
    {
        Motorway,
        Trunk,
        Primary,
        Secondary,
        Tertiary,
        Unclassified,
        Residential,
        ServiceOther
    }
}
